test/extensions/filters/network/zookeeper_proxy/zookeeper_filter_test.cc

```diff
@@ -0,0 +1,264 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_decoder.h"
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_filter.h"
+
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/test_base.h"
+
+#include "gmock/gmock.h"
+
+using testing::NiceMock;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+class ZooKeeperFilterTest : public TestBase {
+public:
+  ZooKeeperFilterTest() { ENVOY_LOG_MISC(info, "test"); }
+
+  void initialize() {
+    config_ = std::make_shared<ZooKeeperFilterConfig>(stat_prefix_, scope_);
+    filter_ = std::make_unique<ZooKeeperFilter>(config_);
+    filter_->initializeReadFilterCallbacks(filter_callbacks_);
+  }
+
+  Buffer::OwnedImpl* encodeConnect(const bool readonly = false, const uint64_t zxid = 100,
+                                   const uint32_t session_timeout = 10,
+                                   const uint32_t session_id = 200,
+                                   const std::string& passwd = "") const {
+    Buffer::OwnedImpl* buffer = new Buffer::OwnedImpl();
+    const uint32_t message_size = readonly ? 28 + passwd.length() + 1 : 28 + passwd.length();
+
+    buffer->writeBEInt<uint32_t>(message_size);
+    buffer->writeBEInt<uint32_t>(0); // Protocol version.
+    buffer->writeBEInt<uint64_t>(zxid);
+    buffer->writeBEInt<uint32_t>(session_timeout);
+    buffer->writeBEInt<uint64_t>(session_id);
+    buffer->writeBEInt<uint32_t>(passwd.length());
+    buffer->add(passwd);
+
+    if (readonly) {
+      char readonly_flag = 0b1;
+      buffer->add(std::string(1, readonly_flag));
+    }
+
+    return buffer;
+  }
+
+  Buffer::OwnedImpl* encodeBadMessage() const {
```

Comment:
Can you not return this by value? IIUC it should be moveable

Reply:
Done.